### ISSUE TITLE
Platforms boot package overhaul

### ISSUE THREAD

**Issue by achilleas-k** [osbuild/images#118]
The boot and firmware package selection is a bit of a mess.  We've known it for a while, but a lot of issues came up concretely in #79.
We need to sort through all the packages and make sure that things are defined where they make the most sense:

Note that I'll use the term "boot packages" to refer to both firmware and other boot-related packages like bootloaders and dracut modules.
Also, I refer to `ostree-` image types as an umbrella term for `iot-` and `edge-` image types.

- Boot packages should be defined on Platform objects.
  - The base architecture platforms should include boot packages that apply to all image types of that architecture.
  - Packages should be added conditionally if they depend on options (e.g. `biosdevname`, UEFI vs BIOS).
- OSTree payload carriers (we should come up with a name for these) might need more thought:
  - The Anaconda ostree installer needs certain packages to boot, but will set up the boot environment on the target machine when it's run.
  - The simplified (CoreOS) installer is similar, but the pipeline also builds a raw image to deliver, so there are two boot configurations being set up.  The machine that the installer and the raw image will run on is the same, so there will be a lot in common, but we will want to have some differences between the two.
  - The ostree container is simpler, but also interesting.  The platform for this image type is always shared with the ostree commit, because that's the payload we're building, but the actual artifact is a container.
  - Based on the previous two points, should we have a way to "inherit" the Platform configuration for one image type while building a different image type with a different platform that includes the first image?  For example, the `ostree-simplified-installer` can have its own platform but also be able to read the `ostree-raw-image` platform configuration for the raw image pipelines (we already reuse the pipeline generation code).

Side issue: Should we rename the `FirmwarePackages` field of the `Platform` to just `Packages` since we also include boot-related packages that are not, strictly speaking, _firmware_?

_Note: Let's ignore the issue of distro-specific package names for now._

**Comment by teg** [comment]
Another potential improvement (I think my initial implementation/proposal for this was not really the best):

Currently we have fairly generic platform packages that we instantiate with lots of options for each image type in the distro (for instance whether or not UEFI is supported etc).

I think it would make for conceptually simpler to understand code if we made more specific platform types/objects, so that from the distro package you could do:
`platform.NewEC2X86()`, `platform.NewMetalX86()`, `platform.NewRPi()`, `platform.NewKVMX86()` (or maybe this should be libvirt/openstack,... specific).

I think the interesting one is `Metal`: this would be, I think a generic bare metal platform where we don't make assumptions about what hardware is there (anything supported by RHEL I suppose). We could in the future imagine making this configurable by the user, so it would take arguments obtained from the blueprint, if ever we wanted to let the user select what sort of hardware they want support for specifically.

We could also imagine having several of these `Metal` platforms, under different assumptions. In particular we may want `MetalEdge` just in case R4E supports a smaller subset of hardware than generic RHEL (I think only UEFI for instance?) and we therefore want to make it more refined.

**Comment by teg** [comment]
> Should we rename the FirmwarePackages field of the Platform to just Packages since we also include boot-related packages that are not, strictly speaking, firmware?

Makes sense. We might want to structure it as `bootPackages` and `firmwarePackagse` internally (or even have helper types to capture things like `Grub` and `uBoot` etc), but I agree that externally this should just be exposed as `Packages()` as when we call this when building the manifest, we don't care what kinds of packages these are.

**Comment by teg** [comment]
> should we have a way to read the Platform configuration for one image type while building a different image type with a different platform that includes the first image?

I'm not 100% sure I understood the question. Though I agree that the outer and inner platforms could in principle be different. Especially for containers + commits, but even for installers we may want to support a more powerful platform in the target than in the installer itself (maybe the installer does not need firmware for network devices, but the target does, for instance). Not sure how / when we would use that, but as a principle we probably want to always treat the outer and inner platform as separate, even if always instantiated with the same object in practice.

**Comment by achilleas-k** [comment]
> as a principle we probably want to always treat the outer and inner platform as separate, even if always instantiated with the same object in practice.

I think you understood the question fine then even though I phrased it very badly.

As an example, we have
```go
x86_64.addImageTypes(
  &platform.X86{
    BasePlatform: platform.BasePlatform{
      ImageFormat: platform.FORMAT_RAW,
    },
    BIOS:       false,
    UEFIVendor: "fedora",
  },
  iotRawImgType,
)
```
for the raw image type and then
```go
x86_64.addImageTypes(
  &platform.X86{
    BasePlatform: platform.BasePlatform{
      ImageFormat: platform.FORMAT_RAW,
      FirmwarePackages: []string{
        "grub2-efi-x64-cdboot",
        "grub2-pc",
        "grub2-pc-modules",
        "grub2-tools",
        "grub2-tools-extra",
        "grub2-tools-minimal",
        "iwlwifi-dvm-firmware",
        "iwlwifi-mvm-firmware",
        "microcode_ctl",
        "syslinux",
        "syslinux-nonlinux",
      },
    },
    BIOS:       false,
    UEFIVendor: "fedora",
  },
  iotSimplifiedInstallerImgType,
)
```
for the simplified installer.  

_Note: As a reminder or for anyone reading that's not familiar, the simplified installer pulls an ostree commit, deploys it, builds a raw image, and then puts the image on an ISO with the CoreOS installer configured to install it._

In terms of packages, this is fine.  We wont install any packages for the raw image (they come from the commit) and the boot packages we define on the simplified installer `Platform` are for booting that.

But what about other options?  Is there a situation where the simplified-installer might need to configure something for the raw image that's defined on the `Platform` for the raw but is different from the options for the installer itself?

**Comment by achilleas-k** [comment]
> In terms of packages, this is fine. We wont install any packages for the raw image (they come from the commit) and the boot packages we define on the simplified installer `Platform` are for booting that.

Another note: This might change in the near future if (when) we add multi-artifact builds.

**Comment by ondrejbudai** [comment]
One thing to keep in mind: I fully expect IB to build images with [pre-built UKI](https://www.redhat.com/en/blog/rhel-confidential-virtual-machines-azure-technical-deep-dive) (no shim+grub+kernel+initramfs, just shim+UKI) at some point. I wonder how this is related to this `Platform` discussion.

Surely, we can implement images for confidential computing as separate image types, but it would be pretty cool if it was just a blueprint knob. In this case, what would it mean for the code? Would the image type use a different `Platform`, or would `Platform`s be configurable with which bootloader to choose?
